x/sdk/p2p/router.go

```diff
@@ -0,0 +1,224 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package p2p
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"github.com/ava-labs/avalanchego/version"
+)
+
+var (
+	ErrUnregisteredHandler = errors.New("unregistered app protocol")
+	ErrExistingAppProtocol = errors.New("existing app protocol")
+	ErrUnrequestedResponse = errors.New("unrequested response")
+
+	_ common.AppHandler    = (*Router)(nil)
+	_ validators.Connector = (*Router)(nil)
+)
+
+type pendingAppRequest struct {
+	handler    uint8
+	onResponse AppResponseCallback
+}
+
+type pendingCrossChainAppRequest struct {
+	handler    uint8
+	onResponse CrossChainAppResponseCallback
+}
+
+// Router routes incoming application messages to the corresponding registered
+// app handler. App messages must be made using the registered handler's
+// corresponding Client.
+type Router struct {
+	nodeID ids.NodeID
+
+	handlers                     map[uint8]responder
+	pendingAppRequests           map[uint32]pendingAppRequest
+	pendingCrossChainAppRequests map[uint32]pendingCrossChainAppRequest
+	requestID                    uint32
+	peers                        set.Set[ids.NodeID]
+	lock                         sync.RWMutex
+}
+
+func (r *Router) Connected(ctx context.Context, nodeID ids.NodeID, _ *version.Application) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.peers.Add(nodeID)
+	return nil
+}
+
+func (r *Router) Disconnected(ctx context.Context, nodeID ids.NodeID) error {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.peers.Remove(nodeID)
+	return nil
+}
+
+// NewRouter returns a new instance of Router
+func NewRouter() *Router {
+	return &Router{
+		handlers:                     map[uint8]responder{},
+		pendingAppRequests:           make(map[uint32]pendingAppRequest),
+		pendingCrossChainAppRequests: make(map[uint32]pendingCrossChainAppRequest),
+		peers:                        set.Set[ids.NodeID]{},
+	}
+}
+
+// RegisterAppProtocol reserves an identifier for an application protocol and
+// returns a Client that can be used to send messages for the corresponding
+// protocol.
+func (r *Router) RegisterAppProtocol(handlerID uint8, handler Handler, sender common.AppSender) (*Client, error) {
```

Comment:
What is the situation where we would need different senders for each protocol (rather than having a sender for the router?)

Reply:
There shouldn't be one IMO, the only reason I left sender as a param vs as a field in `Router` was because the router doesn't ever use the `sender`. We can move this to a field in router.